### ISSUE TITLE
Not being able to load generator properly

### ISSUE THREAD

**Issue by Wassap124** [vee-upatising/DLSS#1]
Hey there!  
Thanks for repo, it's really helping me out as a student starting out in GANs.

I know this isn't a public library so you don't really need to help me. but I would love your help.  
I'm trying to run this example to see how it works and got to a point in which a model-collapse started occurring (around epoch 400 or so).

So I used the saved `generatorX.h5` files

The file is saving the generator model, so I instantiate it like so

```python
from keras.models import load_model
dcgan.generator = load_model(model_path + '/generator5.h5')
```

And then there's an error because I didn't compile the model.  
*But* - when continuing the training process (whether I compile the model or not), this happennes:

```
1 [D loss: 0.528100, acc.: 93.50%] [G loss: 0.322768]
2 [D loss: 0.140115, acc.: 100.00%] [G loss: 0.000004]
3 [D loss: 0.001052, acc.: 100.00%] [G loss: 0.000000]
4 [D loss: 0.000722, acc.: 100.00%] [G loss: 0.000000]
```

I don't really understand what's going on because both the discriminator and the generator have a really low loss at this point.

NOTE - this also occurs even if I drop the training manually, *before* the model-collapse starts to take place, so that seems to be out of the equation...

IDE is google.colab

**Comment by Wassap124** [comment]
I THINK I solved it.

After loading the generator, the `combined` model needs to be reconfigured with it (same as in the `__init__` func):

```python
dcgan.generator = load_model(model_path + '/generator5.h5')

# The generator takes low resolution images as input and generates high resolution images
z = Input(shape = dcgan.latent_dim)
img = dcgan.generator(z)

# The discriminator takes generated images as input and determines validity
valid = dcgan.discriminator(img)

# The combined model  (stacked generator and discriminator)
# Trains the generator to fool the discriminator
dcgan.combined = Model(z, valid)
dcgan.combined.compile(loss='binary_crossentropy', optimizer=dcgan.optimizer)

```

**Comment by vee-upatising** [comment]
Hey thanks for reaching out,

I was a recent student that just graduated from college, so I don't mind
helping out other students.

What's happening here is that you are using the code in a way that I didn't
really intend for it to be used (which is fine). But basically what's going
on here is that you are saving the generator every few epochs and it seems
like you are training it to the point of collapse. This is normal, the
training will eventually get to a point where it is no longer learning
anymore. But what you are doing is that you are restarting the training
process with a fully trained generator and a brand new discriminator. This
is what is causing the models to immediately collapse. Because at the start
of training, your generator has already completely outsmarted your
discriminator so they are no longer learning from each other. The generator
and discriminator need to be trained at the same rate, because if one
outsmarts the other then the system collapses. That is what is happening
here. A 400 epoch generator is competing with a 0 epoch discriminator.

I think the problem that you have is that your model is collapsing before
you are getting results that you are satisfied with. So you want to restart
the training of the generator because it is not yet satisfactory correct? I
think that there are multiple things you can do here.
1. Redesign the generator and discriminator. The way I have it designed now
is not guaranteed to work best with all datasets. Try using a different
amount of Conv2D layers. Try using a different amount of filters at each
layer. Try using a different kernel size.
2. Use a different dataset. Once you play around with GANs more you will
get a feel of what type of datasets work and what type of datasets don't
work. It is possible that the dataset you are using is not diverse enough
or doesn't have enough samples to train beyond 400 epochs.

Basically, once your model has reached a point of collapse, that is "as far
as you can go" with that model. Trying to continue the training is
generally not the way you want to go. If your model can only train for 400
epochs before collapsing, then you want to tackle the problem from the
start, which is the way you designed the model and the dataset that you
choose. The key to success and developing intuition is to just experiment.
Don't take the way I designed it as fact, it was specific design for a
specific dataset that I was working with. The point of my repo is to
provide a framework where you can tweak and design your own models with
your own datasets and analyze your results. Have fun with it and just keep
trying new ideas, it's hard to come up with a completely balanced GAN on
the first try.

It's also possible that you don't need to train beyond 400 epochs. In this
example, my model was already producing good results at 80 epochs. Although
the number of epochs is entirely dependent on the dataset you are using so
I can't say any general statements about that.
[image: epoch_80.png]

Hope this helps.
